File: Fin/mstStack_Algo.py
import pickle
import networkx as nx
from queue import PriorityQueue

# Load the saved graph
with open("random_power_graph_50.pkl", "rb") as f:
    G = pickle.load(f)

demand = 5000

# Global trackers
total_power = 0
operation_count = 0

# Priority source order (cleanest to dirtiest)
priority_sources = ["Solar", "Wind", "Hydro", "Coal"]

# Comparator logic using tuples: (clean_score, -power_output)
def build_priority_queues(graph):
    global operation_count
    queues = {src: PriorityQueue() for src in priority_sources}
    for node_id, attr in graph.nodes(data=True):
        tup = (attr['clean_score'], -attr['power_output'], node_id)
        queues[attr['energy_source']].put(tup)
        operation_count += 1  # one operation per queue insertion
    return queues

# Select nodes to meet demand
def select_nodes(queues, demand, graph):
    global total_power, operation_count
    total_power = 0
    selected = []
    for src in priority_sources:
        q = queues[src]
        while not q.empty() and total_power < demand:
            _, _, node_id = q.get()
            selected.append(node_id)
            total_power += graph.nodes[node_id]['power_output']
            operation_count += 1  # one operation per node selected
        if total_power >= demand:
            break
    return selected

# Build MST using Kruskal's algorithm (via networkx)
def build_mst(graph, selected_nodes):
    global operation_count
    subgraph = graph.subgraph(selected_nodes).copy()
    operation_count += len(subgraph.nodes()) + len(subgraph.edges())  # copying overhead

    for u, v, d in subgraph.edges(data=True):
        d['weight'] += d['weight'] * 0.02
        operation_count += 1  # edge weight adjustment

    mst = nx.minimum_spanning_tree(subgraph, weight='weight')
    operation_count += len(mst.edges())  # edge decisions in MST

    total_cost = sum(d['weight'] for u, v, d in mst.edges(data=True))
    operation_count += len(mst.edges())  # summing costs
    return mst, total_cost

# Display selected node info
def display_selected(graph, selected):
    global operation_count
    print("\nSelected nodes for meeting demand:\n")
    for node in selected:
        attr = graph.nodes[node]
        #print(f"Node: {attr['name']}, Source: {attr['energy_source']}, Power: {attr['power_output']}")
        operation_count += 1

    print(f"\nTotal Selected Power: {total_power}\n")

    breakdown = {'Solar': 0, 'Wind': 0, 'Hydro': 0, 'Coal': 0}
    for node in selected:
        src = graph.nodes[node]['energy_source']
        if src in breakdown:
            breakdown[src] += 1
            operation_count += 1

    print(f"Energy Breakdown: {breakdown}")

# Save selected subgraph to a new pickle file
def save_selected_subgraph(graph, selected_nodes, filename="selected_power_graph_50.pkl"):
    subgraph = graph.subgraph(selected_nodes).copy()
    with open(filename, "wb") as f:
        pickle.dump(subgraph, f)
    print(f"\n Subset graph with {len(subgraph.nodes)} nodes saved to '{filename}'.")

# Main routine
def run_clean_power_selection(demand):
    global operation_count
    operation_count = 0  # Reset counter

    queues = build_priority_queues(G)
    selected = select_nodes(queues, demand, G)
    display_selected(G, selected)

    if total_power < demand:
        print("Insufficient power available to meet demand.")
        return

    save_selected_subgraph(G, selected)

    mst, cost = build_mst(G, selected)
    print("\nMST Edges:")
    for u, v, d in mst.edges(data=True):
        print(f"{G.nodes[u]['name']} - {G.nodes[v]['name']} with cost {d['weight']:.2f}")
        operation_count += 1

    print(f"\nTotal MST Cost: {cost:.2f}")
    print(f"\n Total Operation Count: {operation_count}")

# Run the method
run_clean_power_selection(demand)

File: Fin/mstStack_Algo.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <queue>
#include <algorithm>

using namespace std;
int total_power = 0;

// Node structure
struct Node {
    string name;
    string energy_source; // Solar, Wind, Coal, etc.
    int clean_score;       // Lower is better (0 = solar, 5 = coal, etc.)
    int power_output;      // Available power in units
    Node(string n, string e, int c, int p) : name(n), energy_source(e), clean_score(c), power_output(p) {}
    
};

// Edge structure
struct Edge {
    int u, v;
    double weight; // cost of connecting u and v
};


// Disjoint set union aka union-find Kruskal
struct DSU {
    vector<int> parent, rank;
    
    DSU(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    
    int find(int x) {
        if (parent[x] != x) 
            parent[x] = find(parent[x]);
        return parent[x];
    }
    
    bool unite(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX == rootY) return false;
        if (rank[rootX] < rank[rootY]) {
            parent[rootX] = rootY;
        } else if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
        } else {
            parent[rootY] = rootX;
            rank[rootX]++;
        }
        return true;
    }
};


// Graph class
class Graph {
public:
    vector<Node> nodes;
    vector<Edge> edges;
    
    void addEdge(int u, int v, double weight) {
        edges.push_back({u, v, weight});
    }
};

// Comparator for priority queue
struct CompareNodes {
    bool operator()(const Node& a, const Node& b) {
        if (a.clean_score == b.clean_score)
            return a.power_output < b.power_output; 
        return a.clean_score > b.clean_score; 
    }
};

// Priority queues for each energy source
unordered_map<string, priority_queue<Node, vector<Node>, CompareNodes>> stacks;

// Function to populate the stacks
void populateStacks(Graph& g) {
    for (auto& node : g.nodes) {
        stacks[node.energy_source].push(node);
    }
}

// Function to select nodes to meet power demand using priority_queue
vector<Node> selectNodes(unordered_map<string, priority_queue<Node, vector<Node>, CompareNodes>>& stacks, int power_demand) {
    vector<Node> selected;

    vector<string> priority = {"Solar", "Wind", "Hydro", "Coal"}; 

    for (auto& source : priority) {
        while (!stacks[source].empty() && total_power < power_demand) {
            Node node = stacks[source].top(); 
            stacks[source].pop();
            selected.push_back(node);
            total_power += node.power_output;
        }
        if (total_power >= power_demand) break;
    }

    return selected;
}



// display the stacks
void displayStacks() {
    for (auto& pair : stacks) {
        cout << "Energy Source: " << pair.first << endl;
        auto copy = pair.second;
        while (!copy.empty()) {
            Node n = copy.top();
            cout << "  Node: " << n.name << ", Clean Score: " << n.clean_score << ", Power Output: " << n.power_output << endl;
            copy.pop();
        }
        cout << endl;
    }
}

// Function to build MST from selected nodes
void buildMST(Graph& g, vector<Node>& selected_nodes) {
    // Map node name to its index in selected_nodes
    unordered_map<string, int> selected_index;
    for (int i = 0; i < selected_nodes.size(); ++i) {
        selected_index[selected_nodes[i].name] = i;
    }

    vector<Edge> candidate_edges;

    // Only keep edges between selected nodes
    for (auto& edge : g.edges) {
        string u_name = g.nodes[edge.u].name;
        string v_name = g.nodes[edge.v].name;
        if (selected_index.count(u_name) && selected_index.count(v_name)) {
            // Remap u and v to the index inside selected_nodes
            double effective_weight = edge.weight;
            double loss_factor = 0.02;
            effective_weight += (edge.weight * loss_factor);
            candidate_edges.push_back({selected_index[u_name], selected_index[v_name], effective_weight});
        }
    }

    // Sort edges by weight
    sort(candidate_edges.begin(), candidate_edges.end(), [](Edge& a, Edge& b) {
        return a.weight < b.weight;
    });

    // Build MST using Kruskal
    DSU dsu(selected_nodes.size());
    int total_cost = 0;
    cout << "\nMST Edges:\n";
    for (auto& edge : candidate_edges) {
        if (dsu.unite(edge.u, edge.v)) {
            cout << selected_nodes[edge.u].name << " - " << selected_nodes[edge.v].name 
                 << " with cost " << edge.weight << endl;
            total_cost += edge.weight;
        }
    }
    cout << "\nTotal MST Cost: " << total_cost << endl;
}


// Main
int main() {
    Graph g;
    
    // Add nodes 11 nodes (id, energy_source, clean_score, power_output)
    g.nodes.push_back(Node("A", "Solar", 0, 40));
    g.nodes.push_back(Node("B", "Wind", 1, 80));
    g.nodes.push_back(Node("C", "Coal", 5, 150));
    g.nodes.push_back(Node("D", "Solar", 0, 90));
    g.nodes.push_back(Node("E", "Wind", 1, 120));
    g.nodes.push_back(Node("F", "Coal", 5, 200));
    g.nodes.push_back(Node("G", "Solar", 0, 80));
    g.nodes.push_back(Node("H", "Wind", 1, 15));
    g.nodes.push_back(Node("I", "Wind", 1, 30));
    g.nodes.push_back(Node("J", "Wind", 1, 50));
    g.nodes.push_back(Node("K", "Coal", 5, 20));
    

    
    
    // Adding edges (u, v, weight)
    g.addEdge(0, 1, 10);
    g.addEdge(1, 2, 20);
    g.addEdge(0, 3, 15);
    g.addEdge(3, 2, 25);
    g.addEdge(1, 5, 25);
    g.addEdge(4, 7, 25);
    g.addEdge(2, 10, 35);
    g.addEdge(1, 4, 55);
    g.addEdge(2, 8, 25);
    g.addEdge(10, 9, 25);
    g.addEdge(2, 5, 25);

    // Build stacks
    populateStacks(g);

    // Power demand
    int required_power;
    cout << "Enter required power: ";
    cin >> required_power;

    auto selected_nodes = selectNodes(stacks, required_power);

    cout << "Selected nodes for meeting demand:\n\n";
    for (auto& node : selected_nodes) {
        cout << "Node ID: " << node.name
             << ", Source: " << node.energy_source 
             << ", Power: " << node.power_output << endl;
    }

    cout << "\nRequired Power: " << required_power << endl;
    if(required_power > total_power){
        cout << "The grid currently does not have enough power! :(" << endl;
    }

    else{ cout << "Total Power: " << total_power << endl; }

    // Build MST
    buildMST(g, selected_nodes);


    return 0;
}